### PR TITLE
Protocol Cleanup and Refactor

### PR DESCRIPTION
**Pull request by SimonBoothroyd**
## Description
This PR aims to do a general cleanup of the workflow protocols. The main focus will be on making the input / output descriptors more general and extensible.

## TODOS

- [X] Replace the old stub functions with cleaner descriptors
- [X] Remove the need to manually define the private descriptor attribute and value.
- [X] Add automated descriptor docstrings ([e.g](https://property-estimator.readthedocs.io/en/protocol_cleanup/api/generated/propertyestimator.protocols.coordinates.BuildCoordinatesPackmol.html#propertyestimator.protocols.coordinates.BuildCoordinatesPackmol.mass_density))
- [X] Refactor merge behaviour enum to allow future extension.
- [X] Add a new `UNDEFINED` class to distinguish between `None` and unset values.
- [x] Fix existing tests.
- [x] Add per protocol tests

## Status
- [x] Ready to go

### CODE REVIEW

File: propertyestimator/protocols/analysis.py [hunk 1]
```diff
@@ -6,13 +6,14 @@
 from os import path
 
 import numpy as np
+import typing
```
Comment (j-wags) [inline propertyestimator/protocols/analysis.py]: This may be unused

File: propertyestimator/protocols/coordinates.py [hunk 2]
```diff
@@ -13,96 +13,76 @@
...
+    mass_density = protocol_input(
+        docstring='The target density of the created system.',
+        type_hint=unit.Quantity,
+        default_value=0.95 * unit.grams / unit.milliliters
+    )
+
+    box_aspect_ratio = protocol_input(
+        docstring='The aspect ratio of the simulation box.',
+        type_hint=list,
+        default_value=[1.0, 1.0, 1.0]
```
Comment (j-wags) [inline propertyestimator/protocols/coordinates.py]: I'm not 100% sure, but it's not clear whether an object which uses this value would use a _copy_ of this list, or _the list itself_ (in which case modifying one instance of the class could affect another)

File: propertyestimator/protocols/forcefield.py [hunk 3]
```diff
@@ -43,54 +43,34 @@ class WaterModel(Enum):
...
-        self._coordinate_file_path = None
-        self._substance = None
-        self._water_model = BaseBuildSystemProtocol.WaterModel.TIP3P
-
-        # Outputs
-        self._system_path = None
+    force_field_path = protocol_input(
+        docstring='The file path to the force field parameters to assign to the system.',
+        type_hint=str,
+        default_value=UNDEFINED
```
Comment (j-wags) [inline propertyestimator/protocols/forcefield.py]: I like `UNDEFINED` vs. `None` distinction. This has bitten me before, and I may steal the convention.

File: propertyestimator/utils/serialization.py [hunk 4]
```diff
@@ -247,7 +247,7 @@ class TypedJSONEncoder(json.JSONEncoder):
         np.float64: lambda x: {'value': float(x)},
         np.int32: lambda x: {'value': int(x)},
         np.int64: lambda x: {'value': int(x)},
-        np.ndarray: lambda x: {'value': x.tolist()},
+        np.ndarray: lambda x: {'value': x.tolist()}
     }
```
Comment (j-wags) [inline propertyestimator/utils/serialization.py]: Wow. This is great. I need to steal this for serialization in the toolkit.

File: propertyestimator/utils/utils.py [hunk 5]
```diff
@@ -19,7 +19,7 @@ def find_types_with_decorator(class_type, decorator_type):
     ----------
     class_type: class
         The class to pull attributes from.
-    decorator_type: str
+    decorator_type: type
         The type of decorator to search for.
```
Comment (j-wags) [inline propertyestimator/utils/utils.py]: In the new code, is there a difference between `type`s and `class`es? [This SO post](https://stackoverflow.com/a/4479888) seems to say that they're the same thing in python 3.

File: propertyestimator/workflow/decorators.py [hunk 6]
```diff
@@ -1,136 +1,267 @@
...
+    To mark an attribute as an output:
 
-    >>> @protocol_output(value_type=str)
-    >>> def coordinate_file_path(self):
-    >>>     pass
+    >>> from propertyestimator.workflow.protocols import BaseProtocol
+    >>> from propertyestimator.workflow.decorators import protocol_output,
+    >>>
+    >>> class MyProtocol(BaseProtocol):
+    >>>
```
Comment (j-wags) [inline propertyestimator/workflow/decorators.py]: Having a non-indented blank line here might actually break doctests/make this code non-functional

File: propertyestimator/protocols/analysis.py (hunk 1 above)
Comment (SimonBoothroyd) [reply propertyestimator/protocols/analysis.py]: It gets used on line 132 I think!

File: propertyestimator/utils/utils.py (hunk 5 above)
Comment (SimonBoothroyd) [reply propertyestimator/utils/utils.py]: I _think_ the two are the same now although I'm never sure which to use for docstrings - I have swapped the old `class` to `type` in b8a79e30730682cb2478e4d77ac2ffd4b44b4b9e to at least be consistent here!

**Comment by SimonBoothroyd** [comment]
@j-wags thanks for the review! We haven't got doctests set up yet, but if you want to open a PR to do so that would be welcome!

**Pull request merged**